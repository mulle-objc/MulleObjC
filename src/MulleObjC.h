//
//  MulleObjC.h
//  MulleObjC
//
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//

// because standalone versions must define FASTIDs

//#ifdef MULLE_OBJC_RUNTIME_VERSION
//# error "do not include the mulle-objc-runtime before MulleObjC.h"
//#endif

#import "import.h"

// classes
#import "MulleObjCLoader.h"

#import "MulleDynamicObject.h"
#import "MulleObject.h"

#import "NSAutoreleasePool.h"
#import "NSCondition.h"
#import "NSConditionLock.h"
#import "NSInvocation.h"
#import "NSLock.h"
#import "NSMethodSignature.h"
#import "NSNull.h"
#import "NSObject.h"
#import "NSProxy.h"
#import "NSRecursiveLock.h"
#import "NSThread.h"


// protocols and protocolclasses
#import "MulleObjCClassCluster.h"
#import "MulleObjCException.h"
#import "MulleObjCProtocol.h"
#import "MulleObjCRootObject.h"
#import "MulleObjCRuntimeObject.h"
#import "MulleObjCSingleton.h"
#import "MulleObjCTaggedPointer.h"
#import "NSCoding.h"
#import "NSContainer.h"
#import "NSCopying.h"
#import "NSCopyingWithAllocator.h"
#import "NSFastEnumeration.h"
#import "NSLocking.h"
#import "NSObjectProtocol.h"

// categories
#import "NSObject+NSCodingSupport.h"

// structs
// otherwise none.. all C

// functions
#import "MulleObjCAllocation.h"
#import "MulleObjCAutoreleasePool.h"
#import "MulleObjCExceptionHandler.h"
#import "MulleObjCFunctions.h"
#import "MulleObjCHashFunctions.h"
#import "MulleObjCIvar.h"
#import "MulleObjCProperty.h"
#import "MulleObjCPrinting.h"
#import "MulleObjCStackFrame.h"
#import "MulleObjCUniverse.h"
#import "NSByteOrder.h"

