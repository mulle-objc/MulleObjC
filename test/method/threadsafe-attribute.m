#include <MulleObjC/MulleObjC.h>

#include <stdio.h>


@interface Foo


- (void) b _MULLE_OBJC_METHOD_USER_ATTRIBUTE_1;


@end


@implementation Foo



- (void) a MULLE_OBJC_THREADSAFE_METHOD
{
}


- (void) b
{
}



- (void) c _MULLE_OBJC_METHOD_USER_ATTRIBUTE_0 _MULLE_OBJC_METHOD_USER_ATTRIBUTE_4
{
}


@end


static void  print_desc( SEL sel)
{
   struct _mulle_objc_universe     *universe;
   struct _mulle_objc_descriptor   *desc;

   universe = mulle_objc_global_get_universe( __MULLE_OBJC_UNIVERSEID__);
   desc     = _mulle_objc_universe_lookup_descriptor( universe, sel);

   printf( "%s = %d (0x%x)\n",
      _mulle_objc_descriptor_get_name( desc),
      _mulle_objc_descriptor_get_user_bits( desc),
      _mulle_objc_descriptor_get_bits( desc));
}



int  main( void)
{
   print_desc( @selector( a));
   print_desc( @selector( b));
   print_desc( @selector( c));

   return( 0);
}

