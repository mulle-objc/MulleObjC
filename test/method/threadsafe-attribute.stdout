a = 16 (0x8000)
b = 2 (0x1000)
c = 17 (0x8800)
